evm/src/main/java/org/hyperledger/besu/evm/operation/OperationResult.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.evm.operation;
+
+import org.hyperledger.besu.evm.frame.ExceptionalHaltReason;
+
+/** The Operation result. */
+public class OperationResult {
+  /** The Gas cost. */
+  private final long gasCost;
+
+  /** The Halt reason. */
+  private final ExceptionalHaltReason haltReason;
+
+  /** The increment. */
+  private final int pcIncrement;
+
+  /**
+   * Instantiates a new Operation result when due to an ExceptionalHaltReason.
+   *
+   * @param haltReason the halt reason
+   */
+  private OperationResult(final ExceptionalHaltReason haltReason) {
+    this(0L, haltReason);
+  }
+
+  /**
+   * Instantiates a new Operation result when due to an ExceptionalHaltReason with an associated
+   * gasCost increment. Only InsufficientGas has a gas cost increment associated with the reason for
+   * halting, all other halt reasons should use this constructor.
+   *
+   * @param gasCost the gas cost increment that caused the execution to halt
+   * @param haltReason the halt reason
+   */
+  private OperationResult(final long gasCost, final ExceptionalHaltReason haltReason) {
+    this.gasCost = gasCost;
+    this.pcIncrement = 0;
+    this.haltReason = haltReason;
+  }
+
+  /**
+   * Instantiates a new Operation result.
+   *
+   * @param gasCost the gas cost
+   */
+  public OperationResult(final long gasCost) {
+    this(gasCost, 1);
+  }
+
+  /**
+   * Instantiates a new Operation result.
+   *
+   * @param gasCost the gas cost
+   * @param pcIncrement the increment
+   */
+  public OperationResult(final long gasCost, final int pcIncrement) {
+    this.gasCost = gasCost;
+    this.pcIncrement = pcIncrement;
+    this.haltReason = null;
+  }
+
+  /**
+   * Gets gas cost.
+   *
+   * @return the gas cost
+   */
+  public long getGasCost() {
+    return gasCost;
+  }
+
+  /**
+   * Gets halt reason.
+   *
+   * @return the halt reason
+   */
+  public ExceptionalHaltReason getHaltReason() {
+    return haltReason;
+  }
+
+  /**
+   * Gets increment.
+   *
+   * @return the increment
+   */
+  public int getPcIncrement() {
+    return pcIncrement;
+  }
+
+  /**
+   * Returns a halted OperationResult due to ExceptionalHaltReason.INSUFFICIENT_STACK_ITEMS.
+   *
+   * @return halted OperationResult
+   */
+  public static OperationResult underFlow() {
+    return new OperationResult(ExceptionalHaltReason.INSUFFICIENT_STACK_ITEMS);
```

Comment:
There was some inconsistency between cached and non cached instances of `OperationResult` with halt. I decided to not cache them because a halt is a terminating event for the current EVM execution context and caching them also takes memory. I think with not caching we will be better of as I don't think it will impact the EVM considerably.